rfd/0124-plugin-secrets.md

```diff
@@ -0,0 +1,174 @@
+
+```mermaid
+flowchart LR
+  Plugin
+  Okta[Okta Credentials]
+  Slack[Slack Credentials]
+  OpsGenie[OpsGenie Credentials]
+
+  Plugin-->|credentials type okta|Okta
+  Plugin-->|credentials type slack|Slack
+  Plugin-->|credentials type opsgenie|OpsGenie
+```
+
+```yaml
+kind: plugin_static_credentials
+version: v1
+metadata:
+  name: credentials-name
+  labels:
+    teleport.internal/plugin-label: 48398071-9860-4be6-9f29-ca6ca5bc7155 # unique plugin label.
+    teleport.internal/plugin-credentials-type: slack # a unique type associated with the credentials
+spec:
+  credentials:
+    # Only one of these credential types must be defined at a time. All values in this
+    # are string literals and not file locations.
+    api_token: example-token
+    basic_auth:
+      username: example-user
+      password: example-password
+    oauth_client_secret:
+      client_id: example-client-id
+      client_secret: example-client-secret
+```
+
+If a `Plugin` is using oauth, however, it will not need any `PluginStaticCredentials` objects.
+
+### Changes to the `Plugin` object
+
+The `Plugin` object's `Credentials` field will have a new type of credential called
+`PluginStaticCredentialsRefs` that contains a UUID and a list of plugin types.
+
+```yaml
+plugin_static_credentials_refs:
+  label: 48398071-9860-4be6-9f29-ca6ca5bc7155
+  types:
+    - okta
+    - slack
+    - opsgenie
+```
+
+### Access control
+
+`PluginStaticCredentials` can be written to and deleted by `RoleProxy`. At introduction,
+creating credentials is a flow that will only happen through the web UI, so only `RoleProxy`
+is needed. Future work may determine that we want to open up writing through the CLI.
+
+The object will be only readable by the `RoleAdmin` role (with one exception).
+This is because plugins run on the auth server, so it's the only role that needs to be
+able to actually read the static credentials.
+
+The one exception here is Teleport Assist, which runs in the proxy. The proxy will be
+able to read plugin credentials with the name of `openai-default`, which is the name of
+the Assist plugin.
+
+### Lifecycle of the `PluginStaticCredentials` object
+
+#### Creation
+
+A `PluginStaticCredentials` object must be created at the same time that the `Plugin` is created.
+
+#### Deletion
+
+On deletion, the associated `Plugin` will be stopped and related `PluginStaticCredentials`
+objects will be deleted when deleting its associated plugin.
+
+### UX
+
+#### Web UI
+
+The web UI will remain largely the same with respect to entering credentials for specific
+plugins. This will largely depend on the specific application.
+
+### Audit events
+
+A number of new audit events will be created as part of this effort:
+
+| Event Name | Description |
+|------------|-------------|
+| `plugin.credentials.create` | Emitted when plugin credentials have been created. |
+| `plugin.credentials.delete` | Emitted when plugin credentials have been deleted. |
+
+### Security
```

Comment:
I don't see the mention of the credential resource not being cached and the reasoning behind storing it in DynamoDB we discussed in Slack anywhere, unless I missed it. I think we should call it out explicitly in the design.